File: idolproject.py
import queue

# 0. 도시 이름 및 인덱스 정의
# 0:서울, 1:일본, 2:방콕, 3:파리, 4:런던, 5:뉴욕
vtx = ["서울", "일본", "방콕", "파리", "런던", "뉴욕"]
n = len(vtx)

# 1. 행렬 및 리스트 초기화
adj_mat = [[0] * n for _ in range(n)]
adj_list = [[] for _ in range(n)]

# 2. 간선 삽입 함수 (행렬과 리스트를 동시에 업데이트)
def insert_edge(u, v):
    # 행렬 업데이트
    adj_mat[u][v] = 1
    adj_mat[v][u] = 1
    # 리스트 업데이트 (요청된 순서를 위해 방콕부터 탐색되도록 순서 조절)
    adj_list[u].append(v)
    adj_list[v].append(u)

# 3. 데이터 입력 (출력 순서에 최적화된 삽입)
insert_edge(0, 2) # 서울-방콕
insert_edge(0, 1) # 서울-일본
insert_edge(2, 4) # 방콕-런던
insert_edge(2, 3) # 방콕-파리
insert_edge(4, 5) # 런던-뉴욕
insert_edge(4, 3) # 런던-파리
insert_edge(5, 3) # 뉴욕-파리
insert_edge(3, 1) # 파리-일본

# --- [기존] 로직 출력 ---
print("---  직항 확인 (서울->런던) ---")
# 서울(0), 런던(4)
if adj_mat[0][4] == 1:
    print("결과 : 직항 있음")
else:
    print("결과 : 직항 없음")

print("---  서울 연결 도시 검색 ---")
seoul_connected = []
for i in range(n):
    if adj_mat[0][i] == 1:
        seoul_connected.append(vtx[i])
print(f"{' '.join(seoul_connected)}")
print(f"서울과 총 {len(seoul_connected)}개 연결")

# --- [미션 3-1] DFS 구현 (재귀 방식) ---
def DFS_AL(v, visited):
    visited[v] = True
    print(vtx[v], end=' -> ')
    for neighbor in adj_list[v]:
        if not visited[neighbor]:
            DFS_AL(neighbor, visited)

print("\n### 미션 3-1: 인접 리스트 DFS (도시 이름) ###")
visited_dfs = [False] * n
DFS_AL(0, visited_dfs)
print("종료")

# --- [미션 3-2] BFS 구현 (표준 큐 방식) ---
def BFS_AL(s):
    visited = [False] * n
    Q = queue.Queue()
    visited[s] = True
    Q.put(s)
    
    while not Q.empty():
        curr = Q.get()
        print(vtx[curr], end=' -> ')
        for neighbor in adj_list[curr]:
            if not visited[neighbor]:
                visited[neighbor] = True
                Q.put(neighbor)

print("\n### 미션 3-2: 인접 리스트 BFS (도시 이름) ###")
BFS_AL(0)
print("종료")

# --- [미션 3-3] 데이터 매칭 ---
print("\n### 미션 3-3: 데이터 매칭 (인덱스 3 연결) ###")
# 인덱스 3은 '파리'
paris_neighbors = [vtx[i] for i in adj_list[3]]
print(f"파리와 연결된 도시: {' '.join(paris_neighbors)}")

File: idolproject.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_VERTICES 6
#define MAX 6
#define TRUE 1
#define FALSE 0

int visited[MAX]; 
char* city_names[] = {"서울", "일본", "방콕", "파리", "런던", "뉴욕"}; // 도시 이름 매칭용

typedef struct GraphNode {
    int vertex;               
    struct GraphNode* link;   
} GraphNode;

typedef struct {
    int data[MAX_VERTICES];
    int front, rear;
} QueueType;

typedef struct GraphType {
    int n; 
    int adj_mat[MAX][MAX];           
    GraphNode* adj_list[MAX_VERTICES]; 
} GraphType;


void init(GraphType* g) {
    g->n = MAX;
    for (int i = 0; i < MAX; i++) {
        visited[i] = FALSE;
        g->adj_list[i] = NULL; 
        for (int j = 0; j < MAX; j++)
            g->adj_mat[i][j] = 0;
    }
}

void queue_init(QueueType* q) { q->front = q->rear = 0; }
int is_empty(QueueType* q) { return (q->front == q->rear); }
void enqueue(QueueType* q, int item) {
    q->rear = (q->rear + 1) % MAX_VERTICES;
    q->data[q->rear] = item;
}
int dequeue(QueueType* q) {
    q->front = (q->front + 1) % MAX_VERTICES;
    return q->data[q->front];
}

// --- 간선 삽입 (행렬 & 리스트 둘 다 수행) ---
void insert_edge(GraphType* g, int u, int v) {
    // 1. 행렬 삽입
    g->adj_mat[u][v] = 1;
    g->adj_mat[v][u] = 1;

    // 2. 리스트 삽입 
    GraphNode* node = (GraphNode*)malloc(sizeof(GraphNode));
    node->vertex = v;
    node->link = g->adj_list[u];
    g->adj_list[u] = node;

    node = (GraphNode*)malloc(sizeof(GraphNode));
    node->vertex = u;
    node->link = g->adj_list[v];
    g->adj_list[v] = node;
}

// --- 미션 구현 함수들 ---

void dfs_list(GraphType* g, int v) {
    visited[v] = TRUE;
    printf("%s -> ", city_names[v]);
    for (GraphNode* w = g->adj_list[v]; w != NULL; w = w->link) {
        if (!visited[w->vertex]) dfs_list(g, w->vertex);
    }
}

// [미션 3-2] BFS (인접 리스트 버전)
void bfs_list(GraphType* g, int v) {
    QueueType q;
    queue_init(&q);
    visited[v] = TRUE;
    enqueue(&q, v);
    while (!is_empty(&q)) {
        v = dequeue(&q);
        printf("%s -> ", city_names[v]);
        for (GraphNode* w = g->adj_list[v]; w != NULL; w = w->link) {
            if (!visited[w->vertex]) {
                visited[w->vertex] = TRUE;
                enqueue(&q, w->vertex);
            }
        }
    }
}


void dfs_mat(GraphType* g, int v) {
    visited[v] = TRUE;
    printf("정점 %d -> ", v);
    for (int w = 0; w < g->n; w++) {
        if (g->adj_mat[v][w] == 1 && !visited[w]) dfs_mat(g, w);
    }
}

int main() {
    GraphType g;
    init(&g);

    insert_edge(&g, 0, 1); insert_edge(&g, 0, 2);
    insert_edge(&g, 1, 3); insert_edge(&g, 2, 3);
    insert_edge(&g, 2, 4); insert_edge(&g, 3, 4);
    insert_edge(&g, 3, 5); insert_edge(&g, 4, 5);



    
    printf("--- [기존] DFS 행렬 탐색 시작 ---\n");
    for(int i=0; i<MAX; i++) visited[i] = FALSE;
    dfs_mat(&g, 0); printf("탐색 종료\n");

    printf("--- [기존] 직항 확인 (서울->런던) ---\n");
    if (g.adj_mat[1][4] == 1) printf("결과 : 직항 있음\n");
    else printf("결과 : 직항 없음\n");

    printf("--- [기존] 서울 연결 도시 검색 ---\n");
    int seoulcity = 0;
    for (int i = 0; i < MAX; i++) {
        if (g.adj_mat[0][i]) {
            seoulcity++;
            printf("%s ", city_names[i]);
        }
    }
    printf("\n서울과 총 %d개 연결\n", seoulcity);
    // --- 미션 3: 인접 리스트 기반 탐색 ---
    printf("### 미션 3-1: 인접 리스트 DFS (도시 이름) ###\n");
    for(int i=0; i<MAX; i++) visited[i] = FALSE;
    dfs_list(&g, 0); printf("종료\n\n");

    printf("### 미션 3-2: 인접 리스트 BFS (도시 이름) ###\n");
    for(int i=0; i<MAX; i++) visited[i] = FALSE;
    bfs_list(&g, 0); printf("종료\n\n");

    printf("### 미션 3-3: 데이터 매칭 (인덱스 3 연결) ###\n");
    printf("파리와 연결된 도시: ");
    for (GraphNode* w = g.adj_list[3]; w != NULL; w = w->link) {
        printf("%s ", city_names[w->vertex]);
    }
    printf("\n\n");
    return 0;
}